tests/generative/test_query_model.py

```diff
@@ -54,9 +62,54 @@ def query_engines(request):
     }
 
 
+@pytest.fixture(scope="session")
+def recorder():  # pragma: no cover
```

Comment:
I think there are three things that you need to grok for this to make sense. I'm not sure which, if any, have slipped past you so I'll explain all three -- sorry if some of this is obvious.

Reply:
1. Fixtures are functions. Even though it looks as though you're passing the function to the test, really pytest calls the function and then passed the returned value to the test as an argument. So all fixtures are factories.

Reply:
2. This fixture is what pytest calls a [yield fixture](https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended). These exist to allow for setup and clean-up (e.g. creating and tearing down connections) after the test has run.

If an iterator yields (I guess really if it has a `__next__()` method), pytest calls `next()` on it to get the value to pass to the method and then `next()` again after the test has finished. So the code before the `yield` (we don't have any, but see the examples at that link) executes before the test, then the code after the `yield` executes after the test has completed.

I'm using this in a way that pytest probably didn't anticipate -- to add an assertion after the test. And see the following point for why that's useful.

Reply:
3. The fixture itself returns a function. This is a little unusual (many return data or an object of some kind), but it's a known pattern, particularly in combination with a yield fixture where you've set up some state that you want to operate on in the test.

Reply:
One more point for complete clarity.

It generally be more natural to encapsulate all the state within the fixture, like this:

```
@pytest.fixture(scope="session")
def recorder():  # pragma: no cover
    observed_inputs = ObservedInputs()
    
    yield observed_inputs.record

    # ...
```

The only reason we're not doing that here is because we _also_ want to access that state from `conftest.pytest_terminal_summary()`, which is a pytest [hook](https://docs.pytest.org/en/7.1.x/reference/reference.html#hooks) that we use so we can produce output without having it swallowed by pytests [stdout capture](https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html?highlight=stdout#default-stdout-stderr-stdin-capturing-behaviour).

Reply:
Thank you for such a detailed response, @benbc. I've learnt something from each point. I'd like to make a second pass next week, as well as read the linked documentation. However, that need not hold up this PR.